File: frontend/src/services/leadService.ts
//service
import type { Lead, StatusLead } from "../types/Lead";

const API_URL = 'http://localhost:8080/api/leads';

export const leadService = {

//GET: BUSCAR TODOS OS LEADS
    async getAll(): Promise<Lead[]>{
        const response = await fetch(API_URL);
        if(!response.ok){
            throw new Error('Erro ao buscar leads');
        }
        return response.json();
    },

//POST: CRIAR NOVO LEAD
    async create(lead: Omit<Lead, 'id' | 'dataCadastro' | 'dataAtualizacao'>): Promise<Lead> {
        const response = await fetch(API_URL,{
            method: 'POST',
            headers:{
                'Content-Type': 'application/json',
            },
            body: JSON.stringify(lead),
        });

        if (!response.ok){
            throw new Error ('Erro ao criar lead');            
        }        
        return response.json();
    },

//PATCH: ATUALIZAR APENAS O STATUS DO LEAD
    async updateStatus(id: number, status: StatusLead): Promise<Lead> {
        const response = await fetch(`${API_URL}/${id}/status`,{
            method: 'PATCH',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({ status: status }),
        });

        if (!response.ok){
            throw new Error('Erro ao atualizar o status');
        }
        return response.json();
    },
//DELETE: DELETAR UM LEAD
    async delete (id: number): Promise<void>{
        const response = await fetch(`${API_URL}/${id}`, {
        method: 'DELETE'
        });

        if (!response.ok) {
        throw new Error('Erro ao deletar o status');
        }
    },
};

File: mobile/lib/services/LeadService.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/lead.dart';

class LeadService {
  static const String baseUrl = 'http://10.0.2.2:8080/api/leads';
  static Future<List<Lead>> getAll() async {
    try {
      final response = await http.get(Uri.parse(baseUrl));
      if (response.statusCode == 200) {
        List<dynamic> body = jsonDecode(response.body);
        List<Lead> leads = body.map((json) => Lead.fromJson(json)).toList();
        return leads;
      } else {
        throw Exception('Erro ao buscar leads: ${response.statusCode}');
      }
    } catch (error) {
      print('Erro em getAll: $error');
      throw Exception('Erro ao buscar leads: $error');
    }
  }
  static Future<Lead> create(Lead lead) async {
    try {
      Map<String, dynamic> body = lead.toJson();
      final response = await http.post(
        Uri.parse(baseUrl),
        headers: {
          'Content-Type': 'application/json',
        },
        body: jsonEncode(body),
      );
      if (response.statusCode == 201 || response.statusCode == 200) {
        Map<String, dynamic> json = jsonDecode(response.body);
        return Lead.fromJson(json);
      } else {
        throw Exception('Erro ao criar lead: ${response.statusCode}');
      }

    } catch (error) {
      print('Erro em create: $error');
      throw Exception('Erro ao criar lead: $error');
    }
  }

  static Future<void> delete(int id) async {

    try {
      final response = await http.delete(Uri.parse('$baseUrl/$id'));
      if (response.statusCode == 200 || response.statusCode == 204) {
        return;
      } else {
        throw Exception('Erro ao deletar lead: ${response.statusCode}');
      }

    } catch (error) {
      print('Erro em delete: $error');
      throw Exception('Erro ao deletar lead: $error');
    }
  }
}